content-src/components/TopSites/TopSites.js

```diff
@@ -362,8 +384,118 @@ EditTopSites.propTypes = {
 
 const EditTopSitesIntl = injectIntl(EditTopSites);
 
+const TopSiteForm = React.createClass({
+  getDefaultProps() {
+    return {
+      dispatch: () => {},
+      onClose: () => {}
+    };
+  },
+  getInitialState() {
+    return {
+      title: "",
+      url: "",
+      validationError: false
+    };
+  },
+  handleTitleChange(event) {
+    this.resetValidation();
+    this.setState({"title": event.target.value});
+  },
+  handleUrlChange(event) {
+    this.resetValidation();
+    this.setState({"url": event.target.value});
+  },
+  handleCancel() {
+    this.props.onClose();
+  },
+  handleAdd() {
+    if (this.validateForm()) {
+      let url = this.state.url;
+      // If we are missing a protocol, prepend http://
+      if (!url.startsWith("http")) {
```

Comment:
This might now work in some special cases, for example:

* local files (`C:///documents/abc.pdf` or `/Users/Foo/Bar/baz.html`)
* builtin pages (`about:home`, `about:addon` etc)

Perhaps just leave the user's input alone, and let the browser deal with the protocol autocomplete?